# 测试异步代码

```js
function fetchData() {
  return new Promise<unknown>((res, rej) => {
    setTimeout(() => {
      res('peanut butter')
      // rej('error')
    }, 2000)
  })
}
```

## Promise

为你的测试返回一个Promise，则Jest会等待Promise的resove状态 If the promise is rejected, the test will fail.

For example, let's say that fetchData returns a promise that is supposed to resolve to the string 'peanut butter'. 我们可以使用下面的测试代码︰

```js
test('the data is peanut butter', () => {
  return fetchData().then(data => {
    expect(data).toBe('peanut butter');
  });
});
```

## Async/Await

或者，您可以在测试中使用 async 和 await。 写异步测试用例时，可以在传递给test的函数前面加上async。 例如，可以用来测试相同的 fetchData 方案︰


```js
test('the data is peanut butter', async () => {
  const data = await fetchData();
  expect(data).toBe('peanut butter');
});

test('the fetch fails with an error', async () => {
  expect.assertions(1);
  try {
    await fetchData();
  } catch (e) {
    expect(e).toMatch('error');
  }
});
```

你也可以将 async and await和 .resolves or .rejects一起使用。
```js
test('the data is peanut butter', async () => {
  await expect(fetchData()).resolves.toBe('peanut butter');
});

test('the fetch fails with an error', async () => {
  await expect(fetchData()).rejects.toMatch('error');
});
```
